test/inductor/test_torchinductor.py

```diff
@@ -15256,7 +15256,32 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
                 UniformValueConstantFolder(mod).run()
 
             # there are a couple extra tensors created in `insertable_tensor_check`
-            self.assertTrue(max_live_tensors == 3)
+            self.assertEqual(max_live_tensors, 3)
+
+        def test_constant_folding_no_cuda_sync(self):
+            """Test that constant folding runs on CPU to avoid CUDA syncs."""
+            from torch._inductor.fx_passes.joint_graph import UniformValueConstantFolder
+
+            def fn():
+                x = torch.full([10], 5.0, device=GPU_TYPE)
+                y = x + 1
+                z = y * 2
+                return z
+
+            mod = make_fx(fn)()
+
+            # Set sync debug mode to error - will raise if any CUDA sync occurs
+            prev_mode = torch.cuda.get_sync_debug_mode()
+            try:
+                torch.cuda.set_sync_debug_mode("error")
+                # This should not trigger any CUDA syncs since we fold on CPU
+                cf = UniformValueConstantFolder(mod)
+                cf.run()
+            finally:
+                torch.cuda.set_sync_debug_mode(prev_mode)
+
+            # Verify we got the expected folding
+            self.assertEqual(len(cf.node_replacements), 1)
```

Comment:
Would you explain why there is only 1 replacement? Since I think both y and z in the function `fn` are (uniform) constant and should be replaced

Reply:
we discard replacements when all of their uses are known as constants, to avoid the normal Constant Folder which this inherits from using too much memory